### DISCUSSION TITLE
Want to pass CV2 png in ram to pdf stream, how to do it?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#2106] [Ideas]
Below code is fine to write to local disk.
    params = [cv2.IMWRITE_PNG_COMPRESSION, 9]
    cv2.imencode('.png', new, params)[1].tofile( target)

But, I want to use it directly in ram, instead of writing to local file, how to insert above compressed result into pdf ?
        page.insert_image(page.rect, stream=fp.getvalue())

import fitz
import io
from PIL import Image
# ...
img = doc.extractImage(xref)
bio_in = io.BytesIO(img["image"]) # just to create a file-like object
pil = Image.open(bio_in)  # to be used for Pillow 
bio_out = io.BytesIO() # the output file-like
pil.save(bio_out, format="jpeg") # write to it in a space-saving format
imgdoc = fitz.open("jpg", bio_out.getvalue()) # open it as a PyMuPDF document
pdfbytes = imgdoc.convertToPDF() # return the PDF image of that
# ... and so on like in one of the previous posts

**Comment by JorjMcKie** [comment]
I have no experience with CV2. You must find a way to let it write to a general Python file object - as opposed to a filename string.
You then can prepare a file object via ``target = io.BytesIO()`` and tell CV2 to write to ``target``.
Then use the ``bytes`` object ``target.getvalue()``  further on.

**Comment by nissansz** [comment]
There seems no way to do that.
Can only use jpeg, which seems larger than original pdf file size. And clarity may be lower.

**Reply by JorjMcKie** [reply]
Well, blame CV2 then.

**Comment by nissansz** [comment]
Blame Pillow

**Reply by JorjMcKie** [reply]
disagree: a package outputting stuff that is unable to write to memory is **_decades_** behind

**Comment by nissansz** [comment]
is_success, im_buf_arr = cv2.imencode('.png', imgsrc, params)
 fp = io.BytesIO(im_buf_arr)

above code can save to IO,
but after inserting this stream to pdf, it is same result as pillow, super large size. Compression did not work when using io.

**Reply by JorjMcKie** [reply]
ridiculous package!

**Reply by JorjMcKie** [reply]
you might consider to reprocess (copy) the io output to another io with Pillow, this time using "optimize=True".

**Comment by nissansz** [comment]
can you help try to test convert png from io.BytesIO() into pdf? To see whether the file size becomes super large, the only solution seems to use jpeg, which is not as good as png

 new_image.save(fp, "JPEG")  # saving the image in memory

**Reply by JorjMcKie** [reply]
you are using garbage=3,deflate=True?

**Reply by JorjMcKie** [reply]
The following delivers 2 files of identical size:
```python
import fitz,pathlib
doc=fitz.open()
page=doc.new_page()
imgfile="nur-ruhig.jpg"
page.insert_image(page.rect,filename=imgfile)
5
doc.ez_save("file.pdf")
doc.close()
doc=fitz.open()
page=doc.new_page()
page.insert_image(page.rect,stream=pathlib.Path(imgfile).read_bytes())
5
doc.ez_save("stream.pdf")
```

**Comment by nissansz** [comment]
No option used. see below code

                page = output_doc.new_page(width=595,
                                           height=840)  # make page in target Document
                page.insert_image(page.rect, stream=fp.getvalue())

        # doc.delete_pages(0, 1)
        # output_doc.insert_pdf(doc) 

        # output_doc.save(file+"output_file.pdf")
        output_doc.save(pdf)
        output_doc.close()
        doc.close()

**Reply by JorjMcKie** [reply]
Well, give PyMuPDF a chance to compress the stuff by all means!
Reading the documentation, you will find the recommendation to do that, especially for PNG, which may contain an alpha channel.

**Comment by nissansz** [comment]
Any sample code to compress from io?











------------------ 原始邮件 ------------------


    


        发件人: "Jorj X. McKie" ***@***.***>    






    


        发送时间: 2022-12-06 21:30:57    






    


        收件人: pymupdf/PyMuPDF  ***@***.***>     






    


        抄送: nissansz  ***@***.***>,Author  ***@***.***>     






    


        主题: Re: [pymupdf/PyMuPDF] Want to pass CV2 png in ram to pdf stream, howto do it? (Discussion #2106)    

 

Well, give PyMuPDF a chance to compress the stuff by all means!

 Reading the documentation, you will find the recommendation to do that, especially for PNG, which may contain an alpha channel.

**Reply by JorjMcKie** [reply]
What has ``io`` to do with all that?
We are talking about the `Document.save()` parameters.

**Comment by nissansz** [comment]
tried below, size is large than jpeg 95

    output_doc.save(pdf+'_replaced_png.pdf', garbage=3,deflate=True)

**Reply by JorjMcKie** [reply]
but smaller than naked save()?
anyway that is all you can achieve

**Comment by nissansz** [comment]
The strange thing is that
1. first processing with local png files fold, jpeg default 75, pdf size is 3MB.
2. second processing with above pdf to make slight image rect replacement, jpeg default 75, pdf size became 5MB.

**Comment by nissansz** [comment]
I found below the problem may be due to the first step to read image from pdf. After below reading, the image became large than original png before combining pdf. After I read page from combined pdf, then each image became larger, so no matter how I compress, the image still is large from the beginning from the reading step, any method to read image in original size?

    for i in range(0, doc.page_count):
        zoom = 3.5
        mat = fitz.Matrix(zoom, zoom)
        page = doc[i]
        # pix = orig_page.get_pixmap(matrix=mat)
        pix = page.get_pixmap(matrix=mat)
        img = Image.frombytes("RGB", \
                              [pix.width, pix.height], \
                              pix.samples)
        imgsrc = cv2.cvtColor(numpy.asarray(img), cv2.COLOR_RGB2BGR)
